README.md
# Laravel Docs Generator

This package provide smart commands for Swagger / OpenApi annotations to generate documentation

# Installation

You can install package via composer. Add repository to your composer.json

    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/mindz-team/laravel-docs-generator"
        }
    ],

And run

    composer require mindz-team/laravel-docs-generator

Publish config file

    php artisan vendor:publish --provider "L5Swagger\L5SwaggerServiceProvider"
    php artisan vendor:publish  --provider="Mindz\LaravelDocsGenerator\LaravelDocsGeneratorServiceProvider" --tag="config"

## Cookie based auth in SWAGGER

To use swagger based on cookie based auth modify file `resources/views/vendor/l5-swagger/index.blade.php` and set `requestInterceptor` as follows

    requestInterceptor: function (request) {
        request.headers['X-XSRF-TOKEN'] = getCookie('XSRF-TOKEN');
        return request;
    },

and add method

    function getCookie(name) {
        const nameEQ = name + "=";
        const ca = decodeURIComponent(document.cookie).split(';');
        for (let i = 0; i < ca.length; i++) {
            let c = ca[i];
            while (c.charAt(0) == ' ') c = c.substring(1, c.length);
            if (c.indexOf(nameEQ) == 0) return c.substring(nameEQ.length, c.length);
        }
        return null;
    }

# Perquisites

Package is based on [L5-Swagger](https://github.com/DarkaOnLine/L5-Swagger) library. Make sure you are familiar with its
installation and configuration guide.

# Usage

A package contains set of commands that helps to generate swagger annotations and speed up documentation process
radically.

## Init documentation

It generates a file (default. Annotations.php) that initializes swagger documentation.

```shell
php artisan docs-generate:init-documentation
```

Arguments:

`none`

Options:

`--file` - name of file that stores Annotations to initiates swagger. Default is `annotations`

It returns a list of variables with prefilled values that should be pasted into `config/l5-swagger.php` in `constants`
section. It is importatn that if you do not want to use default values you should change them in `contig/l5-swagger.php`
or use constants defined in `.env`

## Crud controller documentation

It Generate swagger documentation for crud controller.

```shell
php artisan docs-generate:crud-controller
```

Arguments:

`name` - Resource name

Options:

`--schema` - Custom schema name for resource (default is based on resource name)<br>
`--tag` - Documentation tag for resource (default is based on resource name)<br>
`--only` - Comma separated action that should only be generated in case your controller does not provide whole CRUD<br>
`--except` - Comma separated action that should **not** be generated in case your controller does not provide whole
CRUD<br>
`--security` - Endpoint security. Default is `bearerAuth`

Command generates ready to go annotations. To complete docs you should fill your schema fill according to example. In case you do not know how to generate annotations please refer to [this](https://github.com/zircote/swagger-php/tree/master/Examples/petstore-3.0/models) document.
If no limitations have been provided action should generate `Index.php`,`Show.php`,`Store.php`,`Update.php`,`Delete.php` files in resource based folder name.

## Single endpoint documentation

Generate swagger documentation for single endpoint

```shell
php docs-generate:endpoint
```

Arguments:

`path` - Endpoint you want to document without prefix (ex. `/users` or `/users/{user}`)

Options:

`--summary` - Short description of endpoint<br>
`--method` - Http methods like: GET,POST,UPDATE,DELETE<br>
`--tag` - Documentation tag for resource (default is based on last path segment)<br>
`--security` - Endpoint security. Default is `bearerAuth`
`--success` - Response code in case of success. Default is `200`

Command generates single file based on `tag` and `path`. Which contains documentation for single endpoint. Dy default this endpoint is schemaless. Custom schema is available only for `POST` and `UPDATE` methods.

Additionally - if you uses parameter in the path like `/users/{user}` the query string parameter will be created automatically with name according to path parameter and default (integer) type


## Fortify endpoints documentation

Generate swagger documentation fortify functionalities from [laravel/fortify](https://github.com/laravel/fortify) package 

```shell
php docs-generate:fortify
```

Arguments:

`none`

Options:

`none`

To determine which endpoints documentation should be published set them in config file `config/docs-generator.php`


# Configuration

You can determine where annotations files should be stored in your file system using config file published during installation `config/docs-generator.php`

# Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

# Security

issue tracker.

# Credits

Author: Roman Szymański [r.szymanski@mindz.it](mailto:r.szymanski@mindz.it)

# License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-docs-generator` will be documented in this file

## master

- fix(): Multipart names of schema without hyphen

## 1.0.1 - 2021-10-14

- Add `docs-generate:fortify` command
- Add possibility to authenticate swagger by cookie auth


## 1.0.0 - 2021-09-23

- Initial version



src/Commands/FortifyDocsCommand.php
<?php

namespace Mindz\LaravelDocsGenerator\Commands;

use Illuminate\Console\Command;
use Illuminate\Support\Facades\Storage;
use Illuminate\Support\Str;

class FortifyDocsCommand extends Command
{
    protected $signature = 'docs-generate:fortify';

    protected $description = 'Generate fortify annotations files';

    public function __construct()
    {
        parent::__construct();
    }

    public function handle()
    {
        $files = config('docs-generator.fortify_endpoints', []);

        $path = config('docs-generator.annotations_path', app_path() . '/Swagger') .'/'. config('docs-generator.actions_directory', 'Actions');

        $client = Storage::createLocalDriver(['root' => $path]);

        collect($files)->each(function ($file) use ($path, $client) {
            $filename = $file . '.php';

            if (file_exists($path . '/' . $filename)) {
                $this->error('File ' . $filename . ' already exists');
                return;
            }

            $fileExploded = explode('/', $file);
            $file = $fileExploded[count($fileExploded) - 1];

            $method = Str::camel($file . 'Stub');
            $client->put($filename, $this->$method());
        });
    }

    protected function loginStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/login",
 *     summary="Authenticates user",
 *     tags={"authenticate"},
 *     operationId="authenticate-login",
 *
 *     @OA\Response(
 *         response=200,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="email",
 *                  example="example@mindz.it",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="password",
 *                  example="",
 *                  type="string"
 *              )
 *         ),
 *     ),
 * )
 */

EOT;
    }

    protected function csrfCookieStub()
    {
        return <<<EOT
<?php
/**
 * @OA\Get(
 *     path="/csrf-cookie",
 *     summary="Initiate session and sets cookies to authorization for all other endpoints",
 *     tags={"authenticate"},
 *     operationId="authenticate-cookie",
 *
 *     @OA\Response(
 *         response=204,
 *         description="No Content",
 *     ),
 *     security={
 *     }
 * )
 */
EOT;
    }

    protected function logoutStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/logout",
 *     summary="Logout user",
 *     tags={"authenticate"},
 *     operationId="authenticate-logout",
 *
 *     @OA\Response(
 *         response=204,
 *         description="No Content",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
*      @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *    }
 * )
 */
EOT;
    }

    protected function forgetPasswordStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/forgot-password",
 *     summary="Send link to resep password",
 *     tags={"password"},
 *     operationId="forgot-password",
 *
 *     @OA\Response(
 *         response=200,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="email",
 *                  example="example@mindz.it",
 *                  type="string"
 *              )
 *         ),
 *     ),
 *     security={}
 * )
 */
EOT;
    }

    protected function resetPasswordStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/reset-password",
 *     summary="Resets user password",
 *     tags={"password"},
 *     operationId="reset-password",
 *
 *     @OA\Response(
 *         response=200,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="email",
 *                  example="example@mindz.it",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="password",
 *                  example="",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="password_confirmation",
 *                  example="",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="token",
 *                  example="",
 *                  type="string"
 *              )
 *         ),
 *     ),
 *     security={}
 * )
 */
EOT;
    }

    protected function registerStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/register",
 *     summary="Register user",
 *     tags={"register"},
 *     operationId="register",
 *
 *     @OA\Response(
 *         response=201,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=404,
 *         description="Not Found",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="name",
 *                  example="John Doe",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="email",
 *                  example="example@mindz.it",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="password",
 *                  example="",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="password_confirmation",
 *                  example="",
 *                  type="string"
 *              )
 *         ),
 *     ),
 * )
 */
EOT;
    }

    protected function showStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Get(
 *     path="/user",
 *     summary="User profile",
 *     tags={"profile"},
 *     operationId="user-profile",
 *
 *     @OA\Response(
 *         response=200,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *     }
 * )
 */
EOT;
    }

    protected function updateStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Put(
 *     path="/user/profile-information",
 *     summary="Update profile",
 *     tags={"profile"},
 *     operationId="update-profile",
 *
 *     @OA\Response(
 *         response=200,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="name",
 *                  example="John Doe",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="email",
 *                  example="example@mindz.it",
 *                  type="string"
 *              ),
 *         ),
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *     }
 * )
 */
EOT;
    }

    protected function updatePasswordStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Put(
 *     path="/user/password",
 *     summary="Update user password",
 *     tags={"profile"},
 *     operationId="update-password",
 *
 *     @OA\Response(
 *         response=204,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="current_password",
 *                  example="",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="password",
 *                  example="",
 *                  type="string"
 *              ),
 *              @OA\Property(
 *                  property="password_confirmation",
 *                  example="",
 *                  type="string"
 *              ),
 *         ),
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *     }
 * )
 */
EOT;
    }

    protected function confirmPasswordStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/user/confirm-password",
 *     summary="User password confirmation",
 *     tags={"profile"},
 *     operationId="confirm-password",
 *
 *     @OA\Response(
 *         response=201,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="password",
 *                  example="",
 *                  type="string"
 *              ),
 *         ),
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *     }
 * )
 */
EOT;
    }

    protected function passwordConfirmationStatusStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Get(
 *     path="/user/confirmed-password-status ",
 *     summary="User password confirmation status",
 *     tags={"profile"},
 *     operationId="confirm-password-status",
 *
 *     @OA\Response(
 *         response=200,
 *         description="Success",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *     }
 * )
 */
EOT;
    }


    protected function sendVerificationEmailStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/email/verification-notification",
 *     summary="Sends message to verify logged user email",
 *     tags={"email-verification"},
 *     operationId="confirm-email",
 *
 *     @OA\Response(
 *         response=202,
 *         description="Accepted",
 *         @OA\MediaType(mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=204,
 *         description="No Content",
 *         @OA\MediaType(mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType(mediaType="application/json")
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *     }
 * )
 */
EOT;
    }


    protected function emailVerificationStub()
    {
        return <<<EOT
<?php

/**
 * @OA\Get(
 *     path="/email/verify/{id}/{hash}",
 *     summary="User email verification",
 *     tags={"email-verification"},
 *     operationId="confirm-password",
 *
 *     @OA\Parameter(
 *         name="id",
 *         in="path",
 *         description="User id",
 *         required=true,
 *         @OA\Schema(
 *             type="integer"
 *         )
 *     ),
 *     @OA\Parameter(
 *         name="hash",
 *         in="path",
 *         description="User email sha1 hash",
 *         required=true,
 *         @OA\Schema(
 *             type="string"
 *         )
 *     ),
 *     @OA\Parameter(
 *         name="expires",
 *         in="query",
 *         description="Expiration timestamp",
 *         required=true,
 *         @OA\Schema(
 *             type="string"
 *         )
 *     ),
 *     @OA\Parameter(
 *         name="signature",
 *         in="query",
 *         description="Signature hash",
 *         required=true,
 *         @OA\Schema(
 *             type="string"
 *         )
 *     ),
 *     @OA\Response(
 *         response=204,
 *         description="No Content",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
*      @OA\Response(
 *         response=403,
 *         description="Forbidden",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *        {"bearerAuth": {}}
 *     }
 * )
 */
EOT;
    }

}


src/Commands/GenerateDocCommand.php
<?php

namespace Mindz\LaravelDocsGenerator\Commands;

use Illuminate\Console\Command;
use Illuminate\Support\Arr;
use Illuminate\Support\Facades\Storage;
use Illuminate\Support\Str;

class GenerateDocCommand extends Command
{
    protected $signature = 'docs-generate:endpoint {path : endpoint without prefix}  {--summary= : Short description }  {--method= : Http methods like: GET,POST,UPDATE,DELETE } {--tag= : Custom endpoints tag } {--security= : Endpoint security. Default is bearerAuth } {--success-response= : Response code in case of success }';

    protected $description = 'Generate swagger documentation for single endpoint';

    private string $path;
    private string $tag;
    private int $successResponse;
    private string $summary;
    private string $method;
    private string $security;

    public function __construct()
    {
        parent::__construct();
    }

    public function handle()
    {
        $this->path = ltrim($this->argument('path'), '/');
        $this->summary = $this->option('summary') ?? 'Example description';
        $this->tag = $this->option('tag') ?? Arr::first($this->explodedPath());
        $this->method = $this->option('method') ?? "GET";

        if (!in_array(Str::lower($this->method), ['get', "post", "patch", "delete"])) {
            $this->error('Allowed methods are get.post,patch and delete');
            return;
        }

        $this->successResponse = $this->option('success-response') ?? 200;
        $this->security = $this->option('security') ?? 'bearerAuth';

        $path = config('docs-generator.annotations_path', app_path() . '/Swagger');
        $client = Storage::createLocalDriver(['root' => $path]);

        $actionPath = config('docs-generator.actions_directory', 'Actions') . '/' . Str::ucfirst(Str::plural($this->tag));

        $filename = $path . '/' . $actionPath . '/' . Str::ucfirst(Arr::last($this->explodedPath())) . '.php';

        if(file_exists($filename)){
            $this->error('File ' .  $actionPath . '/' . Str::ucfirst(Arr::last($this->explodedPath())) . '.php' . ' already exists');
        }

        $client->put($actionPath . '/' . Str::ucfirst(Arr::last($this->explodedPath())) . '.php', $this->customEndpoint());
    }

    private function explodedPath(): array
    {
        return collect(explode('/', $this->path))
            ->reject(fn($segment) => !$segment || Str::startsWith($segment, '{'))
            ->toArray();
    }

    protected function customEndpoint()
    {
        $lastPath = Arr::last($this->explodedPath());
        $method = sprintf("@OA\%s", Str::ucfirst(Str::lower($this->method)));
        $model = Str::ucfirst(Arr::last($this->explodedPath()));

        $requestBody = in_array(Str::lower($this->method), ['post', 'path']) ? <<<EOT
 *     @OA\RequestBody(
 *         description="{$model} model",
 *         required=true,
 *         @OA\JsonContent(
 *              @OA\Property(
 *                  property="example_id",
 *                  example=10,
 *                  type="integer"
 *              )
 *       )
 *     ),
EOT: ' *';

        $pathParameter = $this->parameterInPath();
        $pathParameterSegment = $pathParameter ? <<<EOT
 *
 *     @OA\Parameter(
 *         name="{$pathParameter}",
 *         in="path",
 *         description="{$pathParameter} Id",
 *         required=true,
 *         @OA\Schema(
 *             type="integer"
 *         )
 *     ),
EOT: ' *';

        return <<<EOT
<?php

/**
 * {$method}(
 *     path="/{$this->path}",
 *     summary="{$this->summary}",
 *     tags={"{$this->tag}"},
 *     operationId="custom-{$lastPath}",
{$pathParameterSegment}
{$requestBody}
 *     @OA\Response(
 *         response={$this->successResponse},
 *         description="Success",
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=403,
 *         description="Forbidden",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *         {"{$this->security}": {}}
 *     }
 * )
 */
EOT;
    }

    private function parameterInPath()
    {
        preg_match('/[{](.*)[}]/', $this->path, $matches);
        return Arr::first($matches);
    }
}


src/LaravelDocsGeneratorServiceProvider.php
<?php

namespace Mindz\LaravelDocsGenerator;

use Illuminate\Support\ServiceProvider;
use Mindz\LaravelDocsGenerator\Commands\FortifyDocsCommand;
use Mindz\LaravelDocsGenerator\Commands\GenerateCrudControllerDocsCommand;
use Mindz\LaravelDocsGenerator\Commands\GenerateDocCommand;
use Mindz\LaravelDocsGenerator\Commands\InitDocumentationCommand;

class LaravelDocsGeneratorServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->commands([
                GenerateCrudControllerDocsCommand::class,
                InitDocumentationCommand::class,
                GenerateDocCommand::class,
                FortifyDocsCommand::class
            ]);
        }

        $this->publishes([
            __DIR__ . '/../config/docs-generator.php' => config_path('docs-generator.php'),
        ], 'config');
    }
}


src/Commands/GenerateCrudControllerDocsCommand.php
<?php

namespace Mindz\LaravelDocsGenerator\Commands;

use Illuminate\Console\Command;
use Illuminate\Support\Facades\Storage;
use Illuminate\Support\Str;

class GenerateCrudControllerDocsCommand extends Command
{
    const INDEX_ACTION = 'index';

    const SHOW_ACTION = 'show';

    const STORE_ACTION = 'store';

    const UPDATE_ACTION = 'update';

    const DESTROY_ACTION = 'destroy';

    const ACTIONS = [
        self::INDEX_ACTION,
        self::SHOW_ACTION,
        self::STORE_ACTION,
        self::UPDATE_ACTION,
        self::DESTROY_ACTION,
    ];

    protected $signature = 'docs-generate:crud-controller {name : Resource name }  {--schema= : custom schema name } {--tag= : custom endpoints tag } {--only= : comma separated action that should only be generated} {--except= : comma separated action that should not be generated} {--security= : endpoint security. Default is bearerAuth }';

    protected $description = 'Generate swagger documentation for crud controller';

    private string $resource;
    private string $tag;
    private string $schema;
    private string $security;

    public function __construct()
    {
        parent::__construct();
    }

    public function handle()
    {
        $this->resource = Str::kebab(Str::plural($this->argument('name')));
        $this->tag = $this->option('tag') ?? $this->resource;
        $this->schema = str_replace('-', '', Str::singular(Str::title($this->option('schema') ?? $this->resource)));
        $this->security = $this->option('security') ?? 'bearerAuth';

        $path = config('docs-generator.annotations_path', app_path() . '/Swagger');
        $client = Storage::createLocalDriver(['root' => $path]);

        $actionPath = config('docs-generator.actions_directory', 'Actions') . '/' . Str::ucfirst(Str::camel($this->resource));
        $schemaPath = config('docs-generator.schemas_directory', 'Schemas') . '/' . $this->schema;

        $actions = collect(self::ACTIONS);

        if ($this->option('only')) {
            $allowedActions = explode(',', $this->option('only'));
            $actions = $actions->reject(fn($action) => !in_array($action, $allowedActions));
        }

        if ($this->option('except')) {
            $allowedActions = explode(',', $this->option('except'));
            $actions = $actions->reject(fn($action) => in_array($action, $allowedActions));
        }

        if (file_exists($path . '/' . $schemaPath . '.php')) {
            $this->error('File ' . $schemaPath . '.php' . ' already exists');
        }

        $client->put($schemaPath . '.php', $this->schemaStub());

        $actions->transform(fn($action) => $action . 'Stub')
            ->each(function ($method) use ($client, $actionPath, $path) {
                $actionFilename = $actionPath . '/' . Str::ucfirst(Str::remove('Stub', $method)) . '.php';

                if (file_exists($path . '/' . $actionFilename)) {
                    $this->error('File ' . $actionFilename . ' already exists');
                }

                $client->put($actionFilename, $this->$method());
            });
    }

    protected function schemaStub()
    {
        $schemaDirectory = config('docs-generator.schemas_directory', 'Schemas');
        $namespace = sprintf("App\Swagger\%s", $schemaDirectory);
        return <<<EOT
<?php

namespace {$namespace};

/**
 * Class {$this->schema}
 * @OA\Schema(
 *     title="{$this->schema}",
 *     description="{$this->schema} model",
 * )
 */
class {$this->schema}
{
    /**
     * @OA\Property(
     *     format="integer",
     *     description="Id of object",
     *     title="ID",
     *     readOnly=true
     * )
     * @var string
     */
    private \$id;

    /**
     * @OA\Property(
     *     format="string",
     *     description="Name",
     *     title="Name"
     * )
     * @var string
     */
    private \$name;

}
EOT;
    }

    protected function indexStub()
    {
        $resourceTitle = str_replace('-', ' ', Str::title($this->resource));

        return <<<EOT
<?php

/**
 * @OA\Get(
 *     path="/{$this->resource}",
 *     summary="Return list of {$resourceTitle}",
 *     tags={"{$this->tag}"},
 *     operationId="index-{$this->resource}",
 *
 *     @OA\Response(
 *         response=200,
 *         description="Success",
 *         @OA\JsonContent(
 *             ref="#/components/schemas/{$this->schema}"
 *         ),
 *
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=403,
 *         description="Forbidden",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *         {"{$this->security}": {}}
 *     }
 * )
 */
EOT;
    }

    protected function storeStub()
    {
        $resourceSingular = Str::singular(Str::snake($this->resource));
        $resourceTitle = str_replace('-', ' ', Str::title($this->resource));

        return <<<EOT
<?php

/**
 * @OA\Post(
 *     path="/{$this->resource}",
 *     summary="Add new {$resourceTitle}",
 *     tags={"{$this->tag}"},
 *     operationId="store-{$this->resource}",
 *
 *     @OA\Response(
 *         response=201,
 *         description="Created",
 *         @OA\JsonContent(
 *             ref="#/components/schemas/{$this->schema}"
 *         ),
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=403,
 *         description="Forbidden",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *          @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *          @OA\JsonContent(
 *             allOf={
 *                 @OA\Schema( ref="#/components/schemas/{$this->schema}"),
 *             },
 *             required={},
 *         )
 *     ),
 *     security={
 *         {"{$this->security}": {}}
 *     }
 * )
 */

EOT;
    }

    protected function showStub()
    {
        $resourceSingular = Str::singular(Str::snake($this->resource));
        $resourceTitle = str_replace('-', ' ', Str::title($this->resource));

        return <<<EOT
<?php

/**
 * @OA\Get(
 *     path="/{$this->resource}/{{$resourceSingular}}",
 *     summary="Show {$resourceTitle}",
 *     tags={"{$this->tag}"},
 *     operationId="show-{$this->resource}",
 *
 *     @OA\Parameter(
 *         name="{$resourceSingular}",
 *         in="path",
 *         description="{$resourceTitle} id",
 *         required=true,
 *         @OA\Schema(
 *             type="integer"
 *         )
 *     ),
 *
 *     @OA\Response(
 *         response=200,
 *         description="Ok",
 *         @OA\JsonContent(
 *             ref="#/components/schemas/{$this->schema}"
 *         ),
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=403,
 *         description="Forbidden",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *         {"{$this->security}": {}}
 *     }
 * )
 */

EOT;
    }

    protected function updateStub()
    {
        $resourceSingular = Str::singular(Str::snake($this->resource));
        $resourceTitle = str_replace('-', ' ', Str::title($this->resource));

        return <<<EOT
<?php

/**
 * @OA\Patch(
 *     path="/{$this->resource}/{{$resourceSingular}}",
 *     summary="Update {$resourceTitle}",
 *     tags={"{$this->tag}"},
 *     operationId="update-{$this->resource}",
 *
 *     @OA\Parameter(
 *         name="{$resourceSingular}",
 *         in="path",
 *         description="{$resourceTitle} id",
 *         required=true,
 *         @OA\Schema(
 *             type="integer"
 *         )
 *     ),
 *
 *     @OA\Response(
 *         response=200,
 *         description="Ok",
 *         @OA\JsonContent(
 *             ref="#/components/schemas/{$this->schema}"
 *         ),
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=403,
 *         description="Forbidden",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=404,
 *         description="Not Found",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=422,
 *         description="Unprocessable Entity",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\RequestBody(
 *         @OA\JsonContent(
 *             allOf={
 *                 @OA\Schema( ref="#/components/schemas/{$this->schema}"),
 *             },
 *             required={},
 *         )
 *     ),
 *     security={
 *         {"{$this->security}": {}}
 *     }
 * )
 */

EOT;
    }

    protected function destroyStub()
    {
        $resourceSingular = Str::singular(Str::snake($this->resource));
        $resourceTitle = str_replace('-', ' ', Str::title($this->resource));

        return <<<EOT
<?php

/**
 * @OA\Delete (
 *     path="/{$this->resource}/{{$resourceSingular}}",
 *     summary="Delete {$resourceTitle}",
 *     tags={"{$this->tag}"},
 *     operationId="destroy-{$this->resource}",
 *
 *     @OA\Parameter(
 *         name="{$resourceSingular}",
 *         in="path",
 *         description="{$resourceTitle} id",
 *         required=true,
 *         @OA\Schema(
 *             type="integer"
 *         )
 *     ),
 *
 *     @OA\Response(
 *         response=204,
 *         description="No Content",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=401,
 *         description="Unauthenticated",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=403,
 *         description="Forbidden",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     @OA\Response(
 *         response=404,
 *         description="Not Found",
 *         @OA\MediaType( mediaType="application/json")
 *     ),
 *     security={
 *         {"{$this->security}": {}}
 *     }
 * )
 */

EOT;
    }

}
